aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -655,4 +734,42 @@ std::tuple<Tensor &,Tensor &> mode_out(const Tensor& self, Dimname dim, bool kee
   return at::mode_out(values, indices, self, dimname_to_position(self, dim), keepdim);
 }
 
-}} // namespace at::native
+TORCH_IMPL_FUNC(isin_Tensor_Tensor_out) (
+  const Tensor& elements, const Tensor& test_elements, bool assume_unique, bool invert, const Tensor& out
+) {
+  if (elements.numel() == 0) {
+    return;
+  }
+
+  // Heuristic taken from numpy's implementation.
+  // See https://github.com/numpy/numpy/blob/fb215c76967739268de71aa4bda55dd1b062bc2e/numpy/lib/arraysetops.py#L575
+  if (test_elements.numel() < static_cast<int64_t>(
+        10.0f * std::pow(static_cast<double>(elements.numel()), 0.145))) {
+    out.fill_(invert);
+    isin_default_stub(elements.device().type(), elements, test_elements, invert, out);
+  } else {
+    isin_sorting(elements, test_elements, assume_unique, invert, out);
+  }
+}
+
+TORCH_IMPL_FUNC(isin_Tensor_Scalar_out) (
+  const Tensor& elements, const c10::Scalar& test_elements, bool assume_unique, bool invert, const Tensor& out
+) {
+  // redispatch to eq / ne
+  if (invert) {
+    at::ne_out(const_cast<Tensor&>(out), elements, test_elements);
+  } else {
+    at::eq_out(const_cast<Tensor&>(out), elements, test_elements);
+  }
+}
+
+TORCH_IMPL_FUNC(isin_Scalar_Tensor_out) (
+  const c10::Scalar& elements, const Tensor& test_elements, bool assume_unique, bool invert, const Tensor& out
+) {
+  // redispatch
+  at::isin_out(const_cast<Tensor&>(out), wrapped_scalar_tensor(elements, test_elements.device()),
```

Comment:
Maybe it'd be more straightforward to do

```
result = (test_elements == elements).any()
if invert:
  torch.logical_not(result)
```
of course it'd have to use any_out to ensure that it returned a bool tensor, as expected

Reply:
Hm I just remembered why I did it this way- it can be faster to redispatch to the `(tensor, tensor)` variation since the brute force CPU kernel iterates with early stopping instead of materializing the `test_elements == elements` tensor.